### ISSUE TITLE
[API Proposal]: `Convert.ToInt32` overload that takes a `ROS<char>` instead of a `string`

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/runtime#61397]
### Background and motivation

I'm working on reading a tar file header, which contains fields that represent numbers in octal base, saved as ASCII characters.

To convert them to a ten base number, I need to do the following:

```cs
byte[] buffer = new[]{ 0x30, 0x30, 0x30, 0x30, 0x36, 0x34, 0x34, 0x0 }; // 0000644\0
string str = Encoding.ASCII.GetString(buffer.AsSpan()); // "000644"
int tenBaseNumber = Convert.ToInt32(str, fromBase: 8); // 420
```

The third line is inefficient. Ideally, I should be able to pass a the `string` as a `ReadOnlySpan<char>`.

This can be a problem in my use case, because when I open a tar archive, I need to iterate through all the tar entries, and each entry header has multiple octal fields like the one above.

### API Proposal

```cs
namespace System
{
    public static class Convert
    {
        public static int ToInt32(ReadOnlySpan<char> value, int fromBase);
    }
}
```     

### API Usage

```cs
byte[] buffer = new[]{ 0x30, 0x30, 0x30, 0x30, 0x36, 0x34, 0x34, 0x0 }; // 0000644\0
string str = Encoding.ASCII.GetString(buffer.AsSpan()); // "000644"
int tenBaseNumber = Convert.ToInt32(str.AsSpan(), fromBase: 8); // 420
```

**Comment by danmoseley** [comment]
Question is whether this would have broad utility vs. just having your own helper method.

**Comment by jkotas** [comment]
> The third line is inefficient. Ideally, I should be able to pass a the string as a ReadOnlySpan<char>

The second line is inefficient too. Ideally, you would parse the number directly from `ReadOnlySpan<byte>`.

**Comment by davidfowl** [comment]
Don't we already have `Utf8Parser.TryParse` for this scenario?

**Comment by stephentoub** [comment]
> Don't we already have Utf8Parser.TryParse for this scenario?

Carlos wants to read octal.

**Comment by davidfowl** [comment]
I was just looking to see if we supported bases on those APIs and it turns out we don't.

**Comment by stephentoub** [comment]
Yeah.  I'd be inclined to say that if we want to support bases, these should be overloads on Utf8Parser, and if we want something similar for chars, on int and friends.  Convert has some unusual legacy behaviors around parsing.

**Comment by tannergooding** [comment]
> Yeah. I'd be inclined to say that if we want to support bases, these should be overloads on Utf8Parser, and if we want something similar for chars, on int and friends

If we were doing that, I imagine the natural extension from what we have today is:
1) A new `System.Globalization.NumberStyles.OctalNumber` enum member (matching `HexNumber`) for parsing
2) Some new "well known" format specifier (like `x`, `X` matching `hexadecimal`) for formatting

I'd think, at most, this would extend to `2`, `8`, `10`, and `16` (where we already expose `10` and `16`). I'm not convinced we really need support for other "arbitrary" bases as they become increasingly more specialized.

**Comment by GrabYourPitchforks** [comment]
IMO there is still utility to the proposed method. If you have a string like a row in a CSV, having spanified overloads on Convert would obviate calling Substring over and over again. This falls under the umbrella of what I consider general spanification goodness for parsing routines.

**Comment by carlossanlop** [comment]
> Question is whether this would have broad utility vs. just having your own helper method.

@danmoseley I don't have an easy way to create a helper method that avoids passing the string. The main reason why I need to convert the byte array to a string is because the characters in the byte array could be in _any_ encoding, not just ASCII.

---

> The second line is inefficient too. Ideally, you would parse the number directly from `ReadOnlySpan<byte>`.

@jkotas so you would prefer a method that automatically detects the encoding of the chars saved in the byte array?

```cs
public static int ToInt32(ReadOnlySpan<byte> value, int fromBase);
```

Should the user indicate the encoding manually?

```cs
public static int ToInt32(ReadOnlySpan<byte> value, int fromBase, Encoding encoding);
```

---

> I'd be inclined to say that if we want to support bases, these should be overloads on Utf8Parser, and if we want something similar for chars, on int and friends. Convert has some unusual legacy behaviors around parsing.

@stephentoub but what if the string is in an encoding that is not Utf8? Do we have parsers for other encodings? It feels like `Convert` would be a better place because it would be encoding-agnostic.

---

@tannergooding It would be great to add more support to octal numbers. They are often used in the Unix world. For example, the `Mode` attribute is saved in octal (644, 755, etc.).

---

@GrabYourPitchforks I noticed the `BitConverter` class has similar methods that take a byte array. Would that class be a better location than `Convert`? I am not very familiar with the specific differences of the two classes, I was hoping you could enlighten me.

We need to be careful to detect the encoding of the characters stored in the byte array. Is it better to automatically detect the encoding of the chars in the byte array, or should the user manually specify the encoding?

```cs
namespace System
{
    // What if we add them to BitConverter?
    public static class BitConverter
    {
        public static int ToInt32(byte[] value, int startIndex); // Exists
        public static int ToInt32(ReadOnlySpan<byte> value); // Exists
        
        public static int ToInt32(ReadOnlySpan<byte> value, int fromBase); // Should autodetect encoding?
        public static int ToInt32(ReadOnlySpan<byte> value, int fromBase, Encoding encoding); // Should the user manually specify encoding?
    }
}
```

---

I suspect implementing the ROS overload would be relatively simple because we already have it in the form of an internal helper method:

https://github.com/dotnet/runtime/blob/b07b26dd18ae64cd4a6b091dd4a2c87aee773eec/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L2214-L2223

**Comment by jkotas** [comment]
> @jkotas so you would prefer a method that automatically detects the encoding of the chars saved in the byte array?

You cannot really detect the encoding of the chars.

Personally, I think this it too niche need to add methods for in the core library. It is trivial to open code a method tailored for the specific case as needed.

**Comment by GrabYourPitchforks** [comment]
@carlossanlop I don't think `BitConverter` would be appropriate for this. That type is essentially a glorified _reinterpret\_cast\<\>_ helper; it doesn't "parse" the incoming data. Since you need a parser, I think your best bet is to stick to a different container type.

**Comment by eerhardt** [comment]
I just ran into this limitation trying to Span-ify some System.ComponentModel.TypeConverter code in dotnet/runtime:

The following code takes a string and tries to turn it into a number:

https://github.com/dotnet/runtime/blob/64e05d839923b8bd7dded3f23fca6994efb552bf/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs#L50-L69

This is in the base number converter class. Each derived class overrides the `FromString` methods and implements them by calling `Convert.ToXXX`:

https://github.com/dotnet/runtime/blob/64e05d839923b8bd7dded3f23fca6994efb552bf/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Int32Converter.cs#L19-L22

https://github.com/dotnet/runtime/blob/64e05d839923b8bd7dded3f23fca6994efb552bf/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Int64Converter.cs#L19-L22

If we had `Convert.ToXXX(ReadOnlySpan<char> value, int fromBase)` overloads, we could refactor this code to not allocate on the `Trim` and the `Substring` in the base class, and just pass the `ReadOnlySpan<char>` to the derived classes, which would just pass the value through to `Convert`.

**Comment by tannergooding** [comment]
Closing, basically everything except octal is supported for UTF-8 and UTF-16 via the normal numberstyles options and its generally faster, more efficient, and usable in more contexts.

If we really have a need for octal more broadly, a separate proposal not involving Convert can be opened.
